### PR TITLE
Feature/8969 inbound event support in app

### PR DESCRIPTION
**Pull request by RonnyB71**
* Creates an endpoint for receiving events in an app (#47)
* Adds support for subscribing to events in an application (#49)
	* In order to subscribe you need a token for the organization obtained through maskinporten and exchanged to an Altinn token.
* Adds asynchronous status checking of messages delivered through Eformidling using the Event system (#48)

## Related Issue(s)
- #47 
- #48 
- #49 

## Verification
- [ ] **Your** code builds clean without any errors or warnings
- [ ] Manual testing done (required)
- [ ] Relevant automated test added (if you find this hard, leave it and we'll help out)
- [ ] All tests run green

## Documentation
- [ ] User documentation is updated with a separate linked PR in [altinn-studio-docs.](https://github.com/Altinn/altinn-studio-docs) (if applicable)

### CODE REVIEW

File: src/Altinn.App.Api/Controllers/EventsReceiverController.cs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+            {
+                bool eventSuccessfullyProcessed = await eventHandler.ProcessEvent(cloudEvent);
+
+                // A return value of 425 will ensure the event system triggers the retry mecanism
+                return eventSuccessfullyProcessed ? Ok() : new StatusCodeResult(425);
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError("Unable to process event {eventType}. An exception was raised while checking for delivery status of message {messageid}. Exception throw {exceptionMessage}", cloudEvent.Type, cloudEvent.Id, ex.Message);
+                return new StatusCodeResult(500);
```
Comment (tjololo) [inline src/Altinn.App.Api/Controllers/EventsReceiverController.cs:70]: Should the method have a ProducesResponseType(500) attribute as it explicitly return it?

File: src/Altinn.App.Core/EFormidling/EformidlingStartup.cs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        {
+            _appIdentifier = appId;
+            _eventsSubscriptionClient = eventsSubscriptionClient;
+            _logger = logger;
+        }
+
+        ///<inheritDoc/>
+        public async Task StartAsync(CancellationToken cancellationToken)
+        {
+            var eventType = "app.eformidling.reminder.checkinstancestatus";
```
Comment (tjololo) [inline src/Altinn.App.Core/EFormidling/EformidlingStartup.cs]: The event type is defined multiple places, I suspect it always should be the same so maybe it should be a global constant some place instead?

File: src/Altinn.App.Core/EFormidling/Extensions/ServiceCollectionExtensions.cs [hunk 3]
```diff
@@ -12,35 +14,33 @@ namespace Altinn.App.Core.EFormidling.Extensions;
...
-    /// <typeparam name="TM">App specific implementation of <see cref="Altinn.App.Core.EFormidling.Interface.IEFormidlingMetadata"/></typeparam>
+    /// <typeparam name="TM">App specific implementation of <see cref="IEFormidlingMetadata"/></typeparam>
     public static void AddEFormidlingServices<TM>(this IServiceCollection services, IConfiguration configuration) where TM: IEFormidlingMetadata
     {
-        services.AddHttpClient<IEFormidlingClient, Altinn.Common.EFormidlingClient.EFormidlingClient>();
-        services.AddTransient<IEFormidlingReceivers, DefaultEFormidlingReceivers>();
-        services.AddTransient<IEFormidlingService, DefaultEFormidlingService>();
-        services.Configure<Altinn.Common.EFormidlingClient.Configuration.EFormidlingClientSettings>(configuration.GetSection("EFormidlingClientSettings"));
-        services.AddTransient(typeof(IEFormidlingMetadata), typeof(TM));
+        AddEFormidlingServices<TM, DefaultEFormidlingReceivers>(services, configuration);
```
Comment (tjololo) [inline src/Altinn.App.Core/EFormidling/Extensions/ServiceCollectionExtensions.cs:24]: 💯  nice cleanup!

File: src/Altinn.App.Core/EFormidling/Implementation/EformidlingStatusCheckEventHandler.cs [hunk 4]
```diff
@@ -0,0 +1,118 @@
+﻿using Altinn.App.Core.EFormidling.Implementation;
+using Altinn.App.Core.Features;
+using Altinn.App.Core.Interface;
+using Altinn.App.Core.Models;
+using Altinn.Common.EFormidlingClient;
+using Altinn.Common.EFormidlingClient.Models;
+using Microsoft.Extensions.Logging;
+
+namespace Altinn.App.Controllers
```
Comment (tjololo) [inline src/Altinn.App.Core/EFormidling/Implementation/EformidlingStatusCheckEventHandler.cs]: Namespace does not match file location

File: src/Altinn.App.Api/Controllers/ProcessController.cs [hunk 5]
```diff
@@ -195,11 +195,6 @@ public class ProcessController : ControllerBase
                 
                 List<ProcessElement> nextElements = await _flowHydration.NextFollowAndFilterGateways(instance, currentTaskId, false);
 
-                if (nextElements.Count == 0)
-                {
-                    return NotFound("Cannot find any valid process elements that can be reached from current task");
```
Comment (tjololo) [inline src/Altinn.App.Api/Controllers/ProcessController.cs:200]: Remove ProducesResponseType(404) when we don't return 404 on no next elemets any more

File: src/Altinn.App.Core/Interface/IEvents.cs [hunk 6]
```diff
@@ -1,3 +1,4 @@
+using Altinn.App.Core.Infrastructure.Clients.Events;
```
Comment (tjololo) [inline src/Altinn.App.Core/Interface/IEvents.cs]: Unused using?

File: src/Altinn.App.Core/Interface/IEventsSubscription.cs [hunk 7]
```diff
@@ -0,0 +1,15 @@
+using Altinn.App.Core.Infrastructure.Clients.Events;
```
Comment (tjololo) [inline src/Altinn.App.Core/Interface/IEventsSubscription.cs]: Should this be placed in Internal folder?

File: src/Altinn.App.Core/Internal/Events/KeyVaultSecretCodeProvider.cs [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+        /// This 
+        /// </summary>
+        /// <param name="keyVaultClient"></param>
+        public KeyVaultEventSecretCodeProvider(ISecrets keyVaultClient)
+        {
+            _keyVaultClient = keyVaultClient;
+        }
+
+        /// <inheritdoc/>
+        public async Task<string> GetSecretCode()
```
Comment (tjololo) [inline src/Altinn.App.Core/Internal/Events/KeyVaultSecretCodeProvider.cs:25]: Should we create an issue for reloading secret if it changes in keyvault without restarting the application?

File: src/Altinn.App.Core/Internal/Events/UnhandledEventHandler.cs [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+    /// </summary>
+    public class UnhandledEventHandler : IEventHandler
+    {
+        /// <inheritdoc/>
+        public string EventType => "app.events.unhandled";
+
+        /// <inheritdoc/>
+        public Task<bool> ProcessEvent(CloudEvent cloudEvent)
+        {
+            throw new NotImplementedException($"Received unhandled event {cloudEvent.Type} with the following data: {JsonSerializer.Serialize(cloudEvent)}");
```
Comment (tjololo) [inline src/Altinn.App.Core/Internal/Events/UnhandledEventHandler.cs:18]: Will this break the code if we introduce a new event that old applications dont know about?

File: src/Altinn.App.Core/Models/AppIdentifier.cs [hunk 10]
```diff
@@ -22,8 +24,82 @@ public class AppIdentifier
...
+        /// </summary>
+        /// <param name="id">Application id on the form org/app</param>
+        public AppIdentifier(string id)
+        {
+            if (string.IsNullOrEmpty(id))
+            {
+                throw new ArgumentNullException(nameof(id));
+            }
+
+            if (id.ContainsMoreThanOne('/'))
```
Comment (tjololo) [inline src/Altinn.App.Core/Models/AppIdentifier.cs]: Would it be better to implement a method string.ContainsExactlyOne ?

**Review by tjololo — COMMENTED** [review]
Nitpicking as always 🙈

**Review by tjololo — APPROVED** [review]
Cant find anything more, trust your judgement on the comments :)

File: src/Altinn.App.Api/Controllers/EventsReceiverController.cs (hunk 1 above)
Comment (RonnyB71) [reply src/Altinn.App.Api/Controllers/EventsReceiverController.cs:70]: Yes, added. I was thinking about removing the return 500 as the exception will get logged anyway and result in a 500. But wanted a very explicit log statement to help identifying potential issues for now.

File: src/Altinn.App.Core/Internal/Events/UnhandledEventHandler.cs (hunk 9 above)
Comment (RonnyB71) [reply src/Altinn.App.Core/Internal/Events/UnhandledEventHandler.cs:18]: Not sure if I follow. This code will only be invoked if a given application receives an event it hasn't subscribed to.

Comment (tjololo) [reply src/Altinn.App.Core/Internal/Events/UnhandledEventHandler.cs:18]: More likely that I'm not following. The application subscribes to a defined list of events and only them? If so it should be ok to throw a NotImplementedException.

File: src/Altinn.App.Core/Models/AppIdentifier.cs (hunk 10 above)
Comment (RonnyB71) [reply src/Altinn.App.Core/Models/AppIdentifier.cs]: Yep:)

File: src/Altinn.App.Core/EFormidling/EformidlingStartup.cs (hunk 2 above)
Comment (RonnyB71) [reply src/Altinn.App.Core/EFormidling/EformidlingStartup.cs]: Done:)

File: src/Altinn.App.Core/Internal/Events/UnhandledEventHandler.cs (hunk 9 above)
Comment (RonnyB71) [reply src/Altinn.App.Core/Internal/Events/UnhandledEventHandler.cs:18]: Yep! This is basically saying to the Event System that "I haven't implemented support for the event you just sent me!". I case very unlikely to happen though, but at least we then have it explicit if it should happen.

**Pull request merged**